packages/relay-kit/src/packs/safe-4337/types.ts

```diff
@@ -1,19 +1,41 @@
-import Safe from '@safe-global/protocol-kit'
-import { SafeSignature } from '@safe-global/safe-core-sdk-types'
+import Safe, { EthersAdapter } from '@safe-global/protocol-kit'
+import { SafeVersion } from '@safe-global/safe-core-sdk-types'
+import { ethers } from 'ethers'
+import SafeOperation from './SafeOperation'
 
-export type Safe4337Options = {
-  protocolKit: Safe
+type ExistingSafeOptions = {
+  safeAddress: string
+}
+
+type PredictedSafeOptions = {
+  owners: string[]
+  threshold: number
+  safeVersion?: SafeVersion
+  saltNonce?: string
+}
+
+export type Safe4337InitOptions = {
+  ethersAdapter: EthersAdapter
   bundlerUrl: string
-  paymasterUrl?: string
   rpcUrl: string
+  safeModulesVersion?: string
+  customContracts?: {
+    entryPointAddress?: string
+    safe4337ModuleAddress?: string
+    addModulesLibAddress?: string
+  }
+  options: ExistingSafeOptions | PredictedSafeOptions
 }
 
-export interface SafeOperation {
-  readonly data: SafeUserOperation
-  readonly signatures: Map<string, SafeSignature>
-  getSignature(signer: string): SafeSignature | undefined
-  addSignature(signature: SafeSignature): void
-  encodedSignatures(): string
+export type Safe4337Options = {
+  protocolKit: Safe
+  bundlerUrl: string
+  rpcUrl: string
+  bundlerClient: ethers.JsonRpcProvider
+  publicClient: ethers.JsonRpcProvider
```

Comment:
This will make the kit only compatible with Ethers?

Reply:
Internally, yes. Currently, we use ethers throughout the entire SDK. In the future, we may switch to viem once (if :)) we eliminate the dependency.

The `EthersAdapter` is the only compatible option because web3 [lacks](https://github.com/safe-global/safe-core-sdk/blob/d4923af26fa58499ec112039300ad381e61a7542/packages/protocol-kit/src/adapters/web3/Web3Adapter.ts#L276-L319) a `signTypedData` function. We cannot depend on the EthAdapter interface since is doing specific SafeTx or SafeMessage things (not SafeOp). Without adding this logic in the protocol-kit, those functions cannot be used. The current logic resides in the pack

We can refactor the protocol-kit to expose the concrete signTypedData functions instead of the "safe" signTypedData ones. I probably can make it compatible with Web3Adapter by duplicating from the protocol-kit the  necessary logic in the pack and checking the type of the instance passed as param. Currently, it only works with EthersAdapter.